revise process to be jenkins release/official-release driven

jhoblitt opened the pull request.

jhoblitt committed:
rm blurb about branching prior to release process

As this is unimplemented.

jhoblitt committed:
add clo-stubb ref

jhoblitt committed:
make github teams explanation a separate sub-section

jhoblitt committed:
fix typos

jhoblitt committed:
improve rcX < rc1 branch explanation

jhoblitt committed:
add Kicking-off the process sub-sub-sections

jhoblitt committed:
relabel step-by-step instructions as a "manual Release"

jhoblitt committed:
promote subsections -> sections

As this document had 0 top level sections.

jhoblitt committed:
markup fixes

jhoblitt committed:
additional whitespace prior to sections/sub-sections

jhoblitt committed:
revise process to be jenkins release/official-release driven

jhoblitt committed:
simplify conceptual overview

jhoblitt committed:
fix format of git rc tags in clo template

jhoblitt committed:
fix rst code-block errors

gcomoretto commented on index.rst:
```diff
@@ -70,9 +70,6 @@ In gross, the conceptual steps are:
 #. A successful weekly release is identified that forms the basis for the
    official release.
 
-#. (NOT YET IMPLIMENTED) Create a release branch across all EUPS product (DM + external + Aux???)
```
There is a requirements to be able to apply patches on a release. We will need to use branches in that case, however we don't need to create all the branches beforehand. It may be enough to branch just the product that need to be fixed, when required.

gcomoretto commented on index.rst:
```diff
@@ -314,15 +314,30 @@ Branch, Merge, Tag
 Any git repository that needs to be modified for additional ``rc`` should be
-**branched**, have the nessicary changes merged (**TBD**: merge to master and
+**branched** and have the nessicary changes merged to a release branch.  Eg.,
+if changes were needed in ``v42_0_0_rc1`` a branch along the lines of ``v42.x``
```
Probably the branch name should be something like:

b42.0.x

the x may be redundant if we identify a patch branch with a "b" at the beginning instead of a "v". However this can just be cosmetic.
The important aspect if that, if our major is `42.0.0`, the patch will `42.0.1`, therefore the branch shall be identified ad `[v]42.0[.x]`. The procedure shall enforce [.x], or [v] ( or [b] instead), or both.

jhoblitt replied:
I've opened an RFC on moving to semver (https://jira.lsstcorp.org/browse/RFC-546) so this can be nailed down.

gcomoretto replied:
Not sure that the RFC is addressing the branch naming, but just assure that everybody agree to adopt the semantic versioning (I think there should be no problems in that sense).
I still think we shall here use here the proper naming for the branch ``[v]42.0.[x]```.

jhoblitt replied:
Already fixed. This comment is on a stale commit instead of the PR.

gcomoretto commented on index.rst:
```diff
@@ -314,15 +314,30 @@ Branch, Merge, Tag
+**branched** and have the nessicary changes merged to a release branch.  Eg.,
+if changes were needed in ``v42_0_0_rc1`` a branch along the lines of ``v42.x``
+should be created in the repos that need changes. (**TBD**: merge to master and
 cherry-pick to release branch or merge to release branch and merge to
```
This should be expanded. I think, if we have already a rc1, and we want to bring into it a change documented in jira issue DM-XYZ, we should have:

- create a ticket branch ticket/DM-XYZ
- create a release (patch) branch on the corresponding SW package b42.0.x (name tbd)
- pull request from ticket/DM-XYZ branch to b42.0.x release branch
- pull request from ticket/DM-XYZ branch to master (if relevant)

jhoblitt replied:
I've opened an RFC (https://jira.lsstcorp.org/browse/RFC-547) to get dev input on this as it is a dev guide workflow issue.  Do you object to merge this as-is with the TBDs? (Note that this comment is on a stale commit)

gcomoretto replied:
Apart from the mentioned RFC, I think it is better to have the branch naming corrected to ``v42.0.x`` before merging.

jhoblitt replied:
It was fixed last week.  This comment is on a stale commit instead of the PR.

jhoblitt commented on index.rst:
```diff
@@ -70,292 +70,140 @@ In gross, the conceptual steps are:
-Build and Publish ``scipipe`` docker image
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+2nd+ Release Candidate(s)
+-------------------------
```
A blurb about test building an rcX candidate with `stack-os-matrix` prior to running `release/official-release` is needed here.

gcomoretto replied:
This should probably be done by who is in charge of merging the changes into the release branch. 
Maybe this should be automated: as soon as some changes are merged in the release branch, the stack-os-matrix job should be run in order to ensure that the changes are fine.

jhoblitt replied:
A subsubsection on running `stack-os-matrix` prior to attempting to create addition `rc`s has been added.  The merge workflow needs to be part of the dev-guide/RFC'd.

jhoblitt commented on index.rst:
```diff
@@ -70,292 +70,140 @@ In gross, the conceptual steps are:
 
-Branch, Merge, Tag
-^^^^^^^^^^^^^^^^^^
+   SOURCE_GIT_REFS: v42.0.0.rc5 v42.x
```
The precedence here is flipped.  The branch name should be left-most.

gcomoretto replied:
OK

jhoblitt replied:
Fixed.

gcomoretto commented on index.rst:
```diff
@@ -366,75 +214,46 @@ an alphabetic prefix (eg., ``v``).  This has the effect of changing the *eups*
-  ``git`` as the option.
+   SOURCE_GIT_REFS: v42.0.0.rc6
+   RELEASE_GIT_TAG: 42.0.0
+   O_LATEST: false
```
Is this not going to be set to TRUE?

jhoblitt replied:
Not always -- only if it is the highest version numbered release.

jhoblitt replied:
A note discussing `O_LATEST` has been added.

gcomoretto commented on index.rst:
```diff
@@ -366,75 +214,46 @@ an alphabetic prefix (eg., ``v``).  This has the effect of changing the *eups*
 
    git clone https://github.com/lsst/lsst.git
-   git checkout -b 14.0
+   git checkout -b 42.0.0
```
It is confusing calling the branch with the same name of the final release.

We still need to agree on the branch naming convention, as I already pointed out in a previous comment. In this case I suggest something like this:
b42.0
b42.0.x
v42.0.x

jhoblitt replied:
I think I prefer `v42.0.x` out of the prefix/postfix options.

jhoblitt replied:
fixed / made consistent.

gcomoretto commented on index.rst:
```diff
@@ -366,75 +214,46 @@ an alphabetic prefix (eg., ``v``).  This has the effect of changing the *eups*
 
-Branching lsst
-^^^^^^^^^^^^^^^
+Branch ``newinstall.sh`` repo
```
newinstall.sh shall be branched when the first release candidate is created, and added a final official tag when the release is concluded.

jhoblitt replied:
It should be branched after because, at present, the jenkins jobs always use `newinstall.sh` from `master`.

jhoblitt replied:
I'm going to mark this as resolved as the newinstall docker image build and tarball pipeline would need to be modified to look for `newinstall.sh` on a branch for it to be safe to branch in advance.

gcomoretto commented on index.rst:
```diff
@@ -54,7 +54,7 @@
 
 
 The LSST Stack Release Process
```
I think wee should rename it, and avoid the world "Stack".
I would suggest to use "Science Pipelines" instead.

jhoblitt replied:
Agreed

jhoblitt replied:
Done.

gcomoretto commented on index.rst:
```diff
@@ -70,292 +70,140 @@ In gross, the conceptual steps are:
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Any git repository that needs to be modified for additional ``rc`` releases
+should be **branched** and have the necessary changes merged to a release
+branch.  Eg., if changes were needed in ``v42.0.0.rc1`` a "release branch"
```
Until now we used 2 digits release version, for example 16.0.
Do you propose to switch to 3 digits, for example 17.0.0, semantic versioning?

jhoblitt replied:
In the past, 2, 3 and 4 component tags have been used. I would prefer to use `x.y.z` semver but this needs to be RFC'd.

jhoblitt replied:
I'm going to leave this as-is until an RFC is accepted.

gcomoretto commented on index.rst:
```diff
@@ -70,292 +70,140 @@ In gross, the conceptual steps are:
+Any git repository that needs to be modified for additional ``rc`` releases
+should be **branched** and have the necessary changes merged to a release
+branch.  Eg., if changes were needed in ``v42.0.0.rc1`` a "release branch"
+along the lines of ``v42.x`` should be created in the repos that need changes.
```
I would like this to be one of the following:
- b42.0
- b42.0.x
- v42.0.x ( confusing since v usually identify versions)

This naming imply that:
- releases including major and minor changes will require a new branch. 
- releases including only bug fixes are patches made on a release branch created from a major or minor release.

jhoblitt replied:
I vote for `v42.0.x` as it is the smallest change over previous conventions.

jhoblitt replied:
fixed / made consistent.

gcomoretto commented on index.rst:
```diff
@@ -70,292 +70,140 @@ In gross, the conceptual steps are:
+along the lines of ``v42.x`` should be created in the repos that need changes.
 
-https://ci.lsst.codes/blue/organizations/jenkins/sqre%2Finfra%2Fbuild-jupyterlabdemo/activity
+(**TBD**: merge to master and cherry-pick to release branch or merge to release
```
This need to be updated in the dev-guide.
Assuming that all fixes and new functionalities are developed in ticket branches, and by default there is a pull request to master, shall we suggest to have an additional pull request to the release branch?

jhoblitt replied:
The same source branch will only work if there aren't merge conflicts.  I lean slightly towards merge to master and then cherry-picking as it ensure the fix is *always* on master.

jhoblitt replied:
I think the convention, be it forward-port or back-port, should be RFC'd since it for anything non-trivial, sci-pipe devs should be doing the merges so they are responsible for resolving conflicts.

gcomoretto commented on index.rst:
```diff
@@ -70,292 +70,140 @@ In gross, the conceptual steps are:
-    COMPILER: devtoolset-6
-    EUPS_TAG: v42_0_0_rc1
-    RELEASE_IMAGE: lsstsqre/centos:7-stack-lsst_distrib-v42_0_0_rc1
+   SOURCE_GIT_REFS: v42.0.0.rc1 v42.x
```
As we commented today, the branch shall come first.

jhoblitt replied:
fixed

gcomoretto reviewed (changes requested):
The order of the SOURCE_GIT_REFS should be change.
Final release should not have O_LATEST: true?

I think we need to have a chat on:

- branch naming
- when and how to branch newinstall.sh

jhoblitt committed:
fix git ref ordering/precedence

jhoblitt committed:
rename doc to "Science Pipelines Release Process"

jhoblitt committed:
rm blurb about this doc being "unpublished"

jhoblitt committed:
add blurb explaining O_LATEST parameter

jhoblitt committed:
fix speling

jhoblitt committed:
add discussion of testing release branches

with stack-os-matrix prior to attempting to build and publish a new
release candidate.

jhoblitt committed:
normalize format of git ref examples

jhoblitt commented:
Re the non-inline code comment: 

- The silly `SOURCE_GIT_REFS` precedence inverse has been fixed
- a blurb (05d842d) discussing `O_LATEST` was added
- I made an explanation on a code comment as to why `newinstall.sh` should be branched afterwards at this point (essentially, so any fixes merged to master will get picked up).

Do you want to leave this PR open until the RFCs on semver and patch branching are resolved or can it be merged as is?

gcomoretto commented:
I think it should be fine, once the branch naming is fixed.

jhoblitt commented:
Would you mind marking the PR as approved?
